src/language-js/printer-estree.js

```diff
@@ -95,108 +91,52 @@ const { printSwitchCaseConsequent } = require("./print/statement");
-    node.type === "ClassPrivateMethod" ||
-    node.type === "ClassProperty" ||
-    node.type === "PropertyDefinition" ||
-    node.type === "TSAbstractClassProperty" ||
-    node.type === "ClassPrivateProperty" ||
-    node.type === "MethodDefinition" ||
-    node.type === "TSAbstractMethodDefinition" ||
-    node.type === "TSDeclareMethod"
-  ) {
-    // their decorators are handled themselves
-  } else if (
-    isNonEmptyArray(node.decorators) &&
-    // If the parent node is an export declaration and the decorator
-    // was written before the export, the export will be responsible
-    // for printing the decorators.
-    !(
-      parentExportDecl &&
-      locStart(parentExportDecl, { ignoreDecorators: true }) >
-        locStart(node.decorators[0])
-    )
-  ) {
-    const shouldBreak =
-      node.type === "ClassExpression" ||
-      node.type === "ClassDeclaration" ||
-      hasNewlineBetweenOrAfterDecorators(node, options);
-
-    const separator = shouldBreak ? hardline : line;
-
-    path.each((decoratorPath) => {
-      let decorator = decoratorPath.getValue();
-      if (decorator.expression) {
-        decorator = decorator.expression;
-      } else {
-        decorator = decorator.callee;
-      }
-
-      decorators.push(printPath(decoratorPath), separator);
-    }, "decorators");
-
-    if (parentExportDecl) {
-      decorators.unshift(hardline);
-    }
-  } else if (
-    isExportDeclaration(node) &&
-    node.declaration &&
-    isNonEmptyArray(node.declaration.decorators) &&
-    // Only print decorators here if they were written before the export,
-    // otherwise they are printed by the node.declaration
-    locStart(node, { ignoreDecorators: true }) >
-      locStart(node.declaration.decorators[0])
+    type === "ClassMethod" ||
+    type === "ClassPrivateMethod" ||
+    type === "ClassProperty" ||
+    type === "PropertyDefinition" ||
+    type === "TSAbstractClassProperty" ||
+    type === "ClassPrivateProperty" ||
+    type === "MethodDefinition" ||
+    type === "TSAbstractMethodDefinition" ||
+    type === "TSDeclareMethod"
   ) {
-    // Export declarations are responsible for printing any decorators
-    // that logically apply to node.declaration.
-    path.each(
-      (decoratorPath) => {
-        const decorator = decoratorPath.getValue();
-        const prefix = decorator.type === "Decorator" ? "" : "@";
-        decorators.push(prefix, printPath(decoratorPath), hardline);
-      },
-      "declaration",
-      "decorators"
-    );
-  } else {
-    // Nodes with decorators can't have parentheses, so we can avoid
-    // computing pathNeedsParens() except in this case.
-    needsParens = pathNeedsParens(path, options);
+    return linesWithoutParens;
   }
 
-  const parts = [];
+  const decorators = printDecorators(path, options, printPath);
+  // Nodes with decorators can't have parentheses
+  const needsParens = !decorators && pathNeedsParens(path, options);
+
+  const parts = [linesWithoutParens];
```

Comment:
Let's avoid returning an array when possible.

Reply:
What do you mean? It suppose to be a `concat`, even if it doesn't need parens, it can still be a `concat` returned by another function,  why other type? You can use `if (isLabel(doc, ...)) {}` to simplify your check.

Reply:
In general I agree that the check should be more flexible.

> It suppose to be a concat, even if it doesn't need parens

But I'm not sure about this. Why create an array for each node?

Reply:
Fixed.